profiling/src/collections/string_table/mod.rs

```diff
@@ -0,0 +1,279 @@
+    bytes: ChainAllocator<VirtualAllocator>,
+
+    /// The ordered hash set of unique strings. The order becomes the StringId.
+    /// The static lifetime is a lie, it is tied to the `bytes`, which is only
+    /// moved if the string table is moved e.g.
+    /// [StringTable::into_lending_iterator].
+    /// References to the underlying strings should generally not be handed,
+    /// but if they are, they should be bound to the string table's lifetime
+    /// or the lending iterator's lifetime.
+    strings: HashSet<&'static str>,
+}
+
+impl Default for StringTable {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl StringTable {
+    /// Creates a new string table, which initially holds the empty string and
+    /// no others.
+    pub fn new() -> Self {
+        // Keep in mind 32-bit .NET. There is only 2 GiB of virtual memory
+        // total available to an application, and we're not the application,
+        // we're just a piece inside it. Additionally, there may be 2 or more
+        // string tables in memory at a given time. Talk to .NET profiling
+        // engineers before making this any bigger.
+        const SIZE_HINT: usize = 4 * 1024 * 1024;
+        let bytes = ChainAllocator::new_in(SIZE_HINT, VirtualAllocator {});
+
+        let mut strings = HashSet::with_hasher(Hasher::default());
+        // It varies by implementation, but frequently I've noticed that the
+        // capacity after the first insertion is quite small, as in 3. This is
+        // a bit too small and there are frequent reallocations. For one sample
+        // with endpoint + code hotspots, we'd have at least these strings:
+        // - ""
+        // - At least one sample type
+        // - At least one sample unit--already at 3 without any samples.
+        // - "local root span id"
+        // - "span id"
+        // - "trace endpoint"
+        // - A file and/or function name per frame.
+        // So with a capacity like 3, we end up reallocating a bunch on or
+        // before the very first sample. The number here is not fine-tuned,
+        // just skipping some obviously bad, tiny sizes.
+        strings.reserve(32);
+
+        // Always hold the empty string as item 0. Do not insert it via intern
+        // because that will try to allocate zero-bytes from the storage,
+        // which is sketchy.
+        strings.insert("");
+
+        Self { bytes, strings }
+    }
+
+    /// Returns the number of strings currently held in the string table.
+    #[inline]
+    #[allow(clippy::len_without_is_empty)]
+    pub fn len(&self) -> usize {
+        self.strings.len()
+    }
+
+    /// Adds the string to the string table if it isn't present already, and
+    /// returns a [StringId] that corresponds to the order that this string
+    /// was originally inserted.
+    ///
+    /// # Panics
+    /// This panics if the allocator fails to allocate a new chunk/node.
```

Comment:
I know errors are annoying to propagate, but panic will bring down the customer process.

Reply:
Yes, I am aware, but this is just like the `GlobalAlloc` panic'ing if it runs out of memory. Same fringe error condition.